back/flag-service/src/flag/user/spec/UserService.spec-integration.ts

```diff
@@ -0,0 +1,163 @@
+
+      let payload = await new Promise<JwtPayload>((resolve, reject) => jwt.verify(response.jwt, process.env.JWT_SECRET, (err, decoded) => {
+        if (err) reject(err);
+        else resolve(decoded);
+      }));
+      expect(payload).toBeDefined();
+      expect(payload.jti).toBeDefined();
+      expect(payload.iat).toBeDefined();
+      expect(payload.exp).toBeDefined();
+      expect(payload.sub).toBe(response.user._id);
+      expect(payload.userData).toStrictEqual({
+        id: response.user._id,
+        email: response.user.email,
+        nickname: response.user.nickname,
+      });
+    });
+    it('throws an EmailNotFoundError when no user exist with this email in db', () => {
+      expect(userService.login('doesnotexist@example.com', 'thisisasecret123'))
+        .rejects.toThrow(EmailNotFoundError);
+    });
+    it('throws an IncorrectPasswordError when the password is incorrect', () => {
+      expect(userService.login('user@example.com', 'wrongpassword789'))
+        .rejects.toThrow(IncorrectPasswordError);
+    });
+  });
+
+  describe('changePassword', () => {
+    it('changes user\'s password', async () => {
+      let user = await userService.register('user@example.com', 'thisisasecret123', 'jane');
+      let userInDb = await userCollection.findOne({ _id: user._id });
+      expect(await argon2.verify(userInDb.password, 'thisisasecret123')).toBe(true);
+      expect(await argon2.verify(userInDb.password, 'newpassword789')).toBe(false);
+
+      user = await userService.changePassword(user._id, 'thisisasecret123', 'newpassword789');
+      userInDb = await userCollection.findOne({ _id: user._id });
+      expect(await argon2.verify(userInDb.password, 'thisisasecret123')).toBe(false);
+      expect(await argon2.verify(userInDb.password, 'newpassword789')).toBe(true);
+    });
+    it('throws an IncorrectPasswordError when current password is incorrect', async () => {
+      const user = await userService.register('john@example.com', 'thisisasecret123', 'john');
+      expect(userService.changePassword(user._id, 'incorrectpassword456', 'newpassword789'))
+        .rejects.toThrow(IncorrectPasswordError);
+    });
+  });
+
+  describe('changeNickname', () => {
+    it('changes user\'s nickname', async () => {
+      let user = await userService.register('user@example.com', 'thisisasecret123', 'jane');
+      let userInDb = await userCollection.findOne({ _id: user._id });
+      expect(userInDb.nickname).toBe('jane');
+
+      user = await userService.changeNickname(user._id, 'thisisasecret123', 'jane42');
+      userInDb = await userCollection.findOne({ _id: user._id });
+      expect(userInDb.nickname).toBe('jane42');
+    });
+    it('throws an IncorrectPasswordError when password is incorrect', async () => {
+      const user = await userService.register('john@example.com', 'thisisasecret123', 'john');
+      expect(userService.changeNickname(user._id, 'incorrectpassword456', 'john42'))
```

Comment:
does change nickname takes a secret as a second parameter or a password ?

Sidenote, I don't think it should take any of both as parameters, only an ID and the new nickname

Reply:
The reason I went for requiring the password to change nickname is because nicknames are unique values, and therefor changing your nickname may get you in a situation where you can't go back to your old nickname. I take potentially unrevertable changes to your account as serious changes that require security measures.

Moreover, I believe that we should always ask the user's password when they want to make any change to their account. If we so choose to verify the current password in the controller (maybe a guard?) then we shouldn't verify the current password in the service's `changePassword` method either.

Reply:
Okay got it, it's indeed a good idea!

Let's keep the business logic in the service, not in any guard :D